docs/architecture/adr-008-event-subscription.md

```diff
@@ -0,0 +1,178 @@
+      3. A subscription must offer an interface to allow for iteration over
+         incoming events only relevant to that subscription (i.e. it should
+         produce an **event stream**).
+      4. It must be possible to, from outside of the RPC client, receive events
+         from multiple subscriptions' event streams concurrently without
+         interfering with/blocking each other.
+      5. It must be possible to, from outside of the RPC client, handle
+         subscriptions' event streams without blocking other RPC operations.
+   2. Offer the ability to **unsubscribe** from specific queries (i.e. to
+      terminate specific subscriptions).
+2. A pub/sub architecture (with an appropriate concurrency model) that
+   facilitates the subscription interface.
+
+## Decision
+
+### Assumptions
+
+* All blocking operations that deal with I/O must be `async`
+* We will not be ["de-asyncifying" the RPC][issue-318] and will rather, in a
+  future ADR, propose a synchronous architecture as well should we need one.
+
+### Proposed Entities and Relationships
+
+The entities in the diagram below are described in the following subsections.
+The diagram also gives some indication as to the proposed concurrency model for
+the architecture and how the entities relate to each other.
+
+![](assets/rpc-client-subscription-rels.png)
+
+### Subscription
+
+A `Subscription` here is envisaged as an entity that implements the
+[Stream][futures-stream] trait, allowing its owner to asynchronously iterate
+through all of the **relevant** incoming events. Use of such a subscription
+should be as simple as:
+
+```rust
+while let Some(result_event) = subscription.next().await {
+   match result_event {
+      Ok(event) => { /* handle event */ },
+      Err(e) => { /* terminate subscription and report error */ },
+   }
+}
+```
+
+Since the underlying events may not be produced at the same cadence as that of
+the `Subscription` owner's processing, it is proposed here to make use of an
+appropriate type of buffered channel to realize this. The size of this buffer
+must be configurable through the RPC Client.
+
+### RPC Pub/Sub
+
+The RPC Client must asynchronously spawn a component `PubSub` whose
+responsibilities are:
+
+1. Manage subscriptions (subscribe/unsubscribe operations from the RPC Client).
+2. Receive and deserialize incoming `Event`s.
+3. Route `Event`s to `Subscription`s whose `Query` matches the `Event`.
+4. Manage transport-level concerns (i.e. WebSockets connection: keep-alive
+   messages, graceful connection closure, etc.).
+
+Items (3) and (4) could be encapsulated into their own sub-components to allow
+for better separation of concerns and therefore easier testing. For example, (3)
+could be encapsulated into a `PubSubRouter` component and (4) into a component
+that only handles transport-level concerns.
+
+This is an *internal* component whose interface only matters to the RPC Client.
```

Comment:
We should break this down, as this seems to be the core of the new implementation it would help to iron out the details of where concerns are separated. A design for this similar to the high-level overview further up would accompany this greatly.

Reply:
Definitely. I was using last week as a spike to de-risk this, because I wasn't 100% sure how to practically implement this in Rust. I think I have a good idea along these lines, and will hopefully submit a draft PR in the next day or so with an example 👍

Reply:
I'd include the design in this PR, as it is vital for the proposal.

Reply:
Yeah sorry I wasn't clear, I meant to say I'll add the design to this PR ASAP 👍